### PR TITLE
[FIX] sale: no show single novariant attribute in configurator


### PR DESCRIPTION
**Pull request by nle-odoo**
A product with single novariant attribute is normally not shown.

But in the configurator modal it was shown and then added in the product
description when doing exactly the same scenario (add the product) than
without the configurator.

opw-2494331

### CODE REVIEW

**Comment by nle-odoo** [comment]
Hi @awa-odoo 


The issue is that if we use the configurator, a no_create attribute with single value is shown in the configurator and in the description. If we don't use it, it is not shown.

I'm not very used to the configurator code so do you think this fix would be alright and in the right location?

Thanks

**Comment by awa-odoo** [comment]
Hello @nle-odoo 

From what I remember, when the attribute is configured as "never create variant", we *always* want to show it in the product description.
Otherwise there is no way of knowing the value the user selected in the SO, since it's not tied to a product variant.

Now in your use case here, there is a single value, meaning we can infer the information even without the description, but it's a bit of an edge case...
I would say the configurator does what it needs to do: display the value in the description of the product when it's an option configured as "never create".
But if there is no optional products, then we don't show the configurator (there is only 1 product, so nothing to "configure").

I'm afraid that changing this behavior might create other issues with more common use cases.
Especially since the check does not seem very robust (based on a ".d-none" class).

In other words I'm not a fan of this fix, but I don't have a better one to provide just by looking at the code.
I would personally treat this as an unhanded edge case but I'm guessing the client would not be happy with that answer.

Cheers.

**Comment by nle-odoo** [comment]
Yes, I only thought something might need to be done when I saw the order line description changed only when there was an optional product.

We could avoid using d-none by using an explicit class, with something like:

```py
diff --git a/addons/sale/views/variant_templates.xml b/addons/sale/views/variant_templates.xml
index 019c54558cc..6adefed28db 100644
--- a/addons/sale/views/variant_templates.xml
+++ b/addons/sale/views/variant_templates.xml
@@ -5,12 +5,13 @@
         <ul t-attf-class="list-unstyled js_add_cart_variants #{ul_class}" t-att-data-attribute_exclusions="json.dumps(attribute_exclusions)">
             <t t-foreach="product.valid_product_template_attribute_line_ids" t-as="ptal">
                 <!-- Attributes selection is hidden if there is only one value available and it's not a custom value -->
+                <t t-set="single_not_custom" t-value="len(ptal.product_template_value_ids._only_active()) == 1 and not ptal.product_template_value_ids._only_active()[0].is_custom"/>
+                <t t-set="single_and_custom" t-value="len(ptal.product_template_value_ids._only_active()) == 1 and ptal.product_template_value_ids._only_active()[0].is_custom" />
                 <li t-att-data-attribute_id="ptal.attribute_id.id"
                     t-att-data-attribute_name="ptal.attribute_id.name"
-                    t-attf-class="variant_attribute #{'d-none' if len(ptal.product_template_value_ids._only_active()) == 1 and not ptal.product_template_value_ids._only_active()[0].is_custom else ''}">
+                    t-attf-class="variant_attribute #{'d-none' if single_not_custom else ''} #{'single_not_custom' if single_not_custom else ''}">
 
                     <!-- Used to customize layout if the only available attribute value is custom -->
-                    <t t-set="single_and_custom" t-value="len(ptal.product_template_value_ids._only_active()) == 1 and ptal.product_template_value_ids._only_active()[0].is_custom" />
                     <strong t-field="ptal.attribute_id.name" class="attribute_name"/>
 
                     <t t-if="ptal.attribute_id.display_type == 'select'">
````

**Comment by awa-odoo** [comment]
@nle-odoo 

I guess we could have a `data-is_single_and_not_custom` next to the existing `data-is_single_and_custom` yes.

I would look a bit silly since then it's easier to have a separate "is_single" and a separate "is_custom" then checking the combinations but that would make the diff a bit bigger.

I can't see clearly if this would break another use case but please make sure to test it extensively (even if there are some automated testing tours for most common use-cases).

Thanks.

**Comment by nle-odoo** [comment]
Thanks, I added `is_single` since we already have `is_custom` the fix is not too big.

In my tests:

- if we have no_create single attribute: it is not shown and not added in description
- if we have no_create multiple attribute: it is shown and added in description

and the same happen if we have other attributes that are no_create.

**Comment by awa-odoo** [comment]
> Thanks, I added `is_single` since we already have `is_custom` the fix is not too big.
> 
> In my tests:
> 
> * if we have no_create single attribute: it is not shown and not added in description
> * if we have no_create multiple attribute: it is shown and added in description
> 
> and the same happen if we have other attributes that are no_create.
> 
> note: since the fix now is part XML this will only affect people that update or install `sale` after the fix is made available.

Please note that the "variants" template is only the one used in the website_sale application, in the e-commerce, when configuring products.
So it should be tested on a sale order form but *also* on the e-commerce pages to check that it gives the intended behavior.

And finally, I hope this change isn't considered as a regression for people that are already using the product configurator on a product with optional products / other variant values and who would actually need to see the description of ALL the variants values selected (even the ones that only have a single value).

If that's all ok for you, then I guess you can proceed.

**Review by nle-odoo — COMMENTED** [review]
robodoo r+
